default/python/AI/ResourcesAI.py

```diff
@@ -147,24 +141,40 @@ def bake_future_focus(self, pid, focus, update=True, force=False):
             self.baked_planet_info[pid] = self.raw_planet_info.pop(pid)
         return success
 
-    def check_avoid_change_due_to_bureaucracy(self, pinfo, focus, focus_gain, influence_gain) -> bool:
-        if not self.bureaucracy or focus == pinfo.current_focus:
+    def check_avoid_change_due_to_bureaucracy(self, pinfo: PlanetFocusInfo, focus: str) -> bool:
+        """Check whether we better not change a focus to avoid the influence penalty from bureaucracy."""
+        if not self.have_bureaucracy or focus == pinfo.current_focus:
             return False  # No bureaucracy or no change
         last_turn = fo.currentTurn() - 1
         if pinfo.planet.LastTurnColonized == last_turn or pinfo.planet.LastTurnConquered == last_turn:
             return False  # no penalty for newly settled/conquered planets
-        threshold = 3.0
-        if (
-            (focus == INFLUENCE and influence_gain < focus_gain + threshold)
-            or (pinfo.current_focus == INFLUENCE and focus_gain < influence_gain + threshold)
+        if pinfo.current_output[2] + 0.5 < pinfo.possible_output[pinfo.current_focus][2]:
             # avoid repeated focus changes: When influence production already is below target, do not
             # get it further down by switching again (I've seen one planet at -30...)
-            or (pinfo.current_output[2] + 0.5 < pinfo.possible_output[pinfo.current_focus][2])
-        ):
-            debug(f"Avoid focus change on {pinfo.planet.name} due to bureaucracy")
+            debug(f"Avoid focus change on {pinfo.planet.name} due bureaucracy and influence not maxed.")
+            return True
+        threshold = 5.0
+        if self.production_value(pinfo, pinfo.current_focus) + threshold > self.production_value(pinfo, focus):
+            debug(f"Avoid focus change on {pinfo.planet.name} due bureaucracy and gain below {threshold}.")
             return True
         return False
 
+    def production_value(self, pinfo: PlanetFocusInfo, focus: str) -> float:
+        """
+        Calculate value of the production output with the given production focus.
+        Returns 0 of focus is not INDUSTRY, RESEARCH or INFLUENCE, unless focus is pinfo.current_focus
+        """
+        if focus not in (INDUSTRY, RESEARCH, INFLUENCE, pinfo.current_focus):
+            return 0.0
+        # Note that changes in stability may affect other values, so there could be cases where a planet
+        # produces more PP with research focus than with industry focus. This is not correctly evaluated yet,
+        # but let's sum up all three anyway. The whole module should be refactored or possible re-written...
+        return (
```

Comment:
Maybe extract this item to the variables for better readability?

Reply:
The only reason, this is hard to read at the moment, is that the indexes are numbers.
I'd rather make it more readable when doing the big refactoring.